tests/ui/wrong_self_convention2.stderr

```diff
@@ -0,0 +1,11 @@
+error: methods with the following characteristics: (`to_*` and `self` type is not `Copy`) usually take `self` by reference
+  --> $DIR/wrong_self_convention2.rs:28:19
+   |
+LL |         fn to_u64(self) -> u64 {
```

Comment:
This still triggers on the trait implementation, not on the trait definition. Is this intended?

Reply:
Yes, I put that intentionally.
The issue pointed out that a trait method signature is controlled by the trait, and there might be other types implementing the same trait that are not `Copy`.  That's why we should allow a `Copy` type to take `&self` instead of just `self` in the case of `to_*` method when it comes from a trait.  
But, the other way around, where we have a `to_*` method from a trait that requires `self` and the type implementing it is not `Copy` then probably that's not something desirable. As I see (e.g. based on `to_string(&self)` from `std::string::ToString` trait) if you expect types implementing your trait to not always be `Copy` you should mark your `to_*` method as `&self`. That's why I thought a lint, in this case, might be useful. Having said that, I'm not super strong on that, so if someone thinks differently that we should also allow this case, I can change that.

Reply:
Hm, but in order to change this, you would also have to change the trait definition, so you could just as well directly lint on the trait definition. Do you have a reason why it shouldn't lint on trait definitions?

Reply:
One reason is that from trait definition alone you don't know what types will implement it. It might be that a particular trait is implemented only by `Copy` types. E.g. as in the #6727

Reply:
Makes sense. Do you think we should stop linting completely then, if traits are involved?

Reply:
You mean for `to_*` methods or all that `wrong_self_convention` lint supports ?

Reply:
Only for the `to_*` variant, since it has multiple directions it can go, depending on the type being `Copy`. All the other variants should only lint in trait definitions, but not in trait implementations.

Reply:
> All the other variants should only lint in trait definitions, but not in trait implementations.

Yep, that's right.
As for `to_*` variant, I think for sure we shouldn't lint if a `Copy` type implements a `to_*` method taking `&self` from some trait (even though it's `Copy` and could take `self`) - this is what this PR changed. 
Another option, non-`Copy` type implementing trait's `to_*` method taking `self`. I felt that this is not something desirable, as it consumes `self`, so in case of non-`Copy` type the method should name `into_*`.

Reply:
Are you motivated to put some more work into this, so that this lint doesn't trigger on the `to` variant in trait definitions **and** implementations anymore?

Reply:
Yes, sure, I can do that. Do you think there might be some valid use-cases for that scenario ? (non-`Copy` type implementing trait's `to_*` method taking `self`). Or is it just better to stay less stringent?

Reply:
Ok, agree that having this as allow-by-default would be better!

Reply:
PR: https://github.com/rust-lang/rust-clippy/pull/7182